IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -443,44 +567,130 @@ - (MSIDCredentialCacheItem *)tokenWithKey:(__unused MSIDCacheKey *)key
 
 // Read one or more credentials from the keychain that match the key (see credentialItem:matchesKey).
 // If not found, return an empty list without setting an error.
-- (NSArray<MSIDCredentialCacheItem *> *)tokensWithKey:(__unused MSIDCacheKey *)key
-                                           serializer:(__unused id<MSIDCacheItemSerializing>)serializer
-                                              context:(__unused id<MSIDRequestContext>)context
-                                                error:(__unused NSError **)error
+- (NSArray<MSIDCredentialCacheItem *> *)tokensWithKey:(MSIDCacheKey *)key
+                                           serializer:(id<MSIDCacheItemSerializing>)serializer
+                                              context:(id<MSIDRequestContext>)context
+                                                error:(NSError **)error
 {
-    NSArray *tokenItems = [self itemsWithKey:key context:context error:error];
-
-    if (!tokenItems)
+    NSArray<MSIDCredentialCacheItem *> *tokenList = [[NSArray alloc] init];
+    
+    if ([key isKindOfClass:([MSIDDefaultCredentialCacheKey class])])
     {
-        return nil;
+        /*
+         Lazy load app and shared credential blob to sync in memory cache with persistent cache.
+         */
+        [self initializeCredential:key serializer:serializer context:context error:error];
+        
+        if (key.isShared)
+        {
+            tokenList = [self.sharedCredential credentialsWithKey:(MSIDDefaultCredentialCacheKey *)key];
+        }
+        else
+        {
+            tokenList = [self.appCredential credentialsWithKey:(MSIDDefaultCredentialCacheKey *)key];
+        }
     }
+    
+    return tokenList;
+}
 
-    NSMutableArray<MSIDCredentialCacheItem *> *tokenList = [NSMutableArray new];
-    for (NSDictionary *dict in tokenItems)
+- (void)initializeCredential:(MSIDCacheKey *)key
+                  serializer:(id<MSIDCacheItemSerializing>)serializer
+                     context:(id<MSIDRequestContext>)context
+                       error:(NSError **)error
+{
+    if (!key.isShared && ![self.appCredential count])
     {
-        NSData *jsonData = dict[(id)kSecValueData];
-        if (jsonData)
+        MSIDMacCredentialCacheItem *appCredential = [self credentialWithKey:key serializer:serializer context:context error:error];
+        if (appCredential)
         {
-            MSIDCredentialCacheItem *cacheItem = (MSIDCredentialCacheItem *)[serializer deserializeCredentialCacheItem:jsonData];
-            if (cacheItem != nil)
-            {
-                [tokenList addObject:cacheItem];
-            }
-            else
-            {
-                MSID_LOG_WITH_CTX(MSIDLogLevelWarning,context, @"Failed to deserialize account");
-            }
+            self.appCredential = appCredential;
```

Comment:
This sequence of checking [self.appCredential account] and assigning self.appCredential doesn't appear entirely threadsafe. It doesn't seem like we want two threads who call credentialWithKey concurrently to potentially create separate MSIDMacCredentialCacheItem objects. A dispatch_barrier_sync() or similar in initializeCredential might be useful to prevent this for both self.appCredential and self.sharedCredential.

Reply:
Good point. But I would prefer to use the @synchronised directive here as if we use the same concurrent queue which we are using to synchronize keychain calls, it will block keychain operations during this initialization process.